### DISCUSSION TITLE
Not much to show rn, but I decided to make my own implementation.

### DISCUSSION THREAD

**Discussion by MinekPo1** [Its-Jakey/SCPP-old-#5] [Show and tell]
I would love to contribute here, but Java is not a language I am able to work with. I was unable to even build the project.

I'm writhing my implementation in python and I'm going to show it when I have a bit more done!

**Comment by MinekPo1** [comment]
Update: got the lexer working, working on parser.

**Reply by MinekPo1** [reply]
I may have forgotten how hard it is to write a token to AST parser by hand agh.

**Reply by MinekPo1** [reply]
Like not hard, just time consuming.

**Reply by MinekPo1** [reply]
![image](https://user-images.githubusercontent.com/68391527/174900382-2bc42b3b-4b3a-4f81-bf1d-cdd46866ecc2.png)

And I'm not even half way done.

**Reply by MinekPo1** [reply]
Even that is a over statement tbh.

All I got so far is `#include`, `#define`, `if`, `while` and part of `for`

**Reply by MinekPo1** [reply]
The `for` statement alone is 85 lines (finished)

**Reply by MinekPo1** [reply]
All the variable definitions have over a 100 lines
It does not help that I'm getting sleepy and can just accept copilots suggestion for code that looks like someone is payed per line of code.

**Reply by Its-Jakey** [reply]
Does your implementation output SLVM (I renamed the VM) bytecode?

**Reply by Its-Jakey** [reply]
Do you think It might be worth re-writing the entire SCPP variable system and implementing types and structures?

**Reply by MinekPo1** [reply]
> Does your implementation output SLVM (I renamed the VM) bytecode?

Not yet however I plan that.

**Reply by MinekPo1** [reply]
> Do you think It might be worth re-writing the entire SCPP variable system and implementing types and structures?

I think we should use the lax typing of scratch. I was planing to include something alike pythons type hint system in as a side feature included with my impl:

```scpp
//type hint comment:
var a = 123; //type: number
//inline type hint:
var b: string = "hello";
```

This is not checked by the compiler (or interpreter for python's case) but by a static type checker.

However alternatively you could take inspiration from rust, which does type checking at compilation, however the type declarations are alike python.

The final say is up to you, but I would love structures.

Anyway I'm gonna throw some pros and cons for each:

Current:

pros:
- "careless" development
cons:
- "careless" development
- pain do debug
- `("b"+"a"++"a"+"a").lowercase() => "banana"`

Type hints and optional static type checking:

pros:
- almost careless (no quotes) development
- optional
- easy IDE integration
- types which don't exit can be easily added (optional types, unions, ect)
cons:
- since type hints can be omitted, beginers might not use them
- requires more work

Rust like type system:

pros:
- almost careless (no quotes again) development
- often implicit
- types which don't exit can be easily added (optional types, unions, ect)
cons:
- a bit more work
- may confuse beginers

C like type system:

pros:
- thematic
- easy to understand
- easy to implement
cons:
- does not represent underlying system well
- clunky at times
- may be hard to use

VM runtime type checking (combines with any above type systems)

pros:
- intuitive
- easy to debug
cons:
- slow
- hassle to make
- takes away freedom

I also think that memory management would be very nice but I'm gonna create a separate thread for that.

**Reply by MinekPo1** [reply]
Also if you want to add structs, the best way, I think, would be to use an array and bind each member to a index in that array.

**Reply by MinekPo1** [reply]
However doing it like this will be messy with arrays... I guess an implicit pointer or reference would work but that's quite a hack...

**Comment by MinekPo1** [comment]
Ok if you wanna see my spaghetti see [here](https://www.github.com/MinekPo1/PySCPP)

**Reply by Its-Jakey** [reply]
References will be very handy in your implementation!

**Reply by Its-Jakey** [reply]
I saw that you are working on adding parsing for structure definitions. Will structures actually be implemented into the language?

**Reply by MinekPo1** [reply]
> I saw that you are working on adding parsing for structure definitions. Will structures actually be implemented into the language?

Well I dropped them for now, but I might re-add them to **my dialect** of SCPP. If you want to add them to the language then sure. I mainly dropped structs, because I didn't want to deal with all the cases where there would be a pointer to a struct and the compiler would have to detect that and agh.

I have since removed the human right violation that was the old parser, so working on my impl, is less painful so I might revisit that idea. If you like any [differences in my dialect](https://github.com/MinekPo1/PySCPP#non-spec-behaviour), feel free to add them.

**Reply by Its-Jakey** [reply]
Hey sorry about changing the spec so much, just running into some unexplainable bugs like "var array = {1}; print(array[0]) //Printed 2 for some reason"

**Reply by Its-Jakey** [reply]
I'm feeling a little overloaded with my code. The language was redesigned twice over the course of a couple of months before I actually released it, and I don't want to slow you down with your PySCPP by changing the syntax so much. So do whatever you want with the syntax and I'll rewrite the java version with that syntax and rewrite the builtin libraries with that syntax too. For some reason, I just can't keep up with my own code. Because of this, I hand the name "SCPP" to you, and there is no need to keep your syntax like mine or credit me for anything (I'm not being sarcastic BTW), Java's syntax also makes it quite a bit harder to figure out what's going on, and my brain is fried.

**Reply by Its-Jakey** [reply]
forget about that, imma gonna rewrite the language again (with the current syntax, no pointers) and hope that helps.

**Reply by Its-Jakey** [reply]
I'm about 25% done with the new version. I implemented multi-dimensional arrays!

**Reply by Its-Jakey** [reply]
Update: V2.0 is 75% done!
